File: client/web/src/screens/bindings.ts
import type { QuestionRosterEntry } from "../decisions/seam";
import type { BindingDTO, BindingValueDTO, QuestionSwitchDTO } from "../store/protocol";
import type { TaskBindingResult, TaskQuestionSwitchResult } from "../store/store";

// #118's bindings editor, everything decidable: what each binding is called
// in human words, what it currently holds, and whether a draft is worth
// sending. The `.tsx` only threads React state through these — the same
// pure-module split every other `screens/*.ts` uses.
//
// The key vocabulary itself is NOT redefined here. `Core::bindings` returns
// every key it knows (in vocabulary order) plus every row it does not, each
// flagged with `known`; this module only decorates that answer. A TS-side
// list of keys would be a second vocabulary to keep in step with ADR-0015's
// one, and the drift would show up as a row that silently stops rendering.

/** The human name and one-line purpose for a binding this build knows.
 * Looked up by key with a fallback (below), never used as the source of
 * which keys exist. */
export interface BindingCopy {
  label: string;
  help: string;
}

const BINDING_COPY: Record<string, BindingCopy> = {
  "race-series": {
    label: "Race series",
    help: "Which series the next-race answer follows.",
  },
  "trips-calendar": {
    label: "Trips calendar",
    help: "The calendar holding trips — the vacation countdown reads its next event.",
  },
  "city-waste-page": {
    label: "Waste schedule page",
    help: "The council page the collection schedule is read from.",
  },
  // Not a source binding: this one names something the homework pane
  // *draws* rather than something it reads (`bindings.rs`'s own note on the
  // key). It is here rather than in the code because it carries a passcode
  // and this repo is public.
  "homework-link": {
    label: "Homework session link",
    help: "The meeting link the homework answer offers. Kept here, not in the code — it carries a passcode.",
  },
  // Written by the OpenClaw agent (ADR-0032 part 4), not typed here in the
  // ordinary case — the editor still lets the operator hand-correct it.
  "scps-quest": {
    label: "SCPS Photo Quest",
    help: "This month's Photo Quest phrase, as \"YYYY-MM phrase\" — normally set by the agent from forwarded club email.",
  },
};

/** Copy for one binding row. A key this build cannot write still gets a row
 * — it is really in the table — so it falls back to its raw key and says
 * plainly why it is read-only, rather than being dressed up as a binding
 * this build understands. */
export function bindingCopy(binding: BindingDTO): BindingCopy {
  const known = BINDING_COPY[binding.key];
  if (known !== undefined) {
    return known;
  }
  return {
    label: binding.key,
    help: "Set by a newer version of the app. This build can show it, but not change it.",
  };
}

/** What a binding's value should read as on screen. Three states in, three
 * distinct readings out — an unset binding must never render as an empty
 * string beside a set one, which is exactly the collapse the tagged union
 * exists to prevent. */
export function bindingValueLabel(value: BindingValueDTO): string {
  switch (value.state) {
    case "unset":
      return "Not set";
    case "text":
      return value.text;
    case "other":
      return `Not a text value: ${value.raw}`;
  }
}

/** The text a row's input starts at: the current value when it is text, and
 * an empty field otherwise. Deliberately empty for `"other"` too — a value
 * this editor cannot express must not be pre-loaded into a field whose
 * submit would overwrite it with a mangled string. */
export function bindingDraftSeed(value: BindingValueDTO): string {
  return value.state === "text" ? value.text : "";
}

/** Whether two reads of a binding's value are the same fact.
 *
 * The row's input is seeded from the value once and then owned by the
 * typist, so something has to notice when the value underneath it changes —
 * another device's edit arriving on the next pull, or this device's own
 * write confirming. Without it the field keeps showing the old value while
 * the label above it shows the new one, and Save is enabled to push the
 * stale text straight back over what just arrived (#118 review finding). */
export function sameBindingValue(a: BindingValueDTO, b: BindingValueDTO): boolean {
  if (a.state !== b.state) {
    return false;
  }
  if (a.state === "text" && b.state === "text") {
    return a.text === b.text;
  }
  if (a.state === "other" && b.state === "other") {
    return a.raw === b.raw;
  }
  return true;
}

/** The message a failed binding write should show on its own row, or `null`
 * when the last write was someone else's row or succeeded.
 *
 * Every non-`ok` outcome gets words: an enqueue that failed, a core that was
 * busy and a key this build cannot write are all "Save appeared to do
 * nothing" without one. Matched by key so a stale failure from a DIFFERENT
 * binding never bleeds onto this row — the same rule `NowScreen`'s act error
 * follows for item ids. */
export function bindingWriteError(
  lastWrite: TaskBindingResult | null,
  key: string,
): string | null {
  if (lastWrite === null || lastWrite.key !== key || lastWrite.kind === "ok") {
    return null;
  }
  switch (lastWrite.kind) {
    case "unknown_key":
      return "This build doesn't know that binding, so it wasn't saved.";
    case "busy":
      return "The core was busy. Try saving again.";
    case "failed":
      return lastWrite.error ?? "That binding didn't save.";
  }
}

/** Whether a drafted binding is worth sending.
 *
 * Refused here rather than in the core, for the same reason
 * `capture-validation.ts` refuses an empty capture: `Core::set_binding` has
 * no opinion of its own and would enqueue whatever it is handed — and in a
 * table with no DELETE, a binding blanked by a stray keystroke cannot be
 * removed, only overwritten again. A draft equal to what is already stored
 * is also refused: it would be a CAS write whose only effect is a version
 * bump, and every other view would re-render for nothing. */
export function canSubmitBinding(binding: BindingDTO, draft: string): boolean {
  if (!binding.known) {
    return false;
  }
  const trimmed = draft.trim();
  if (trimmed === "") {
    return false;
  }
  return !(binding.value.state === "text" && binding.value.text === trimmed);
}

/** What actually gets sent for a draft — trimmed, so the stored value and
 * the compared-against value can never differ by whitespace the operator
 * cannot see. Call only when [`canSubmitBinding`] passed. */
export function bindingSubmitValue(draft: string): string {
  return draft.trim();
}

// -- the standing-question roster (#714, ADR-0034 decision 4) ---------------
//
// The section stopped being a flat list of keys and became a list of
// *questions* with their keys nested. The relation behind that nesting is
// the core's (`decisions::questions`), reached through
// `questionRosterFromCore`; the two functions below only fold `Core::
// bindings`' answer into it, and take the roster as an argument so they can
// be tested without the wasm seam.

/** One question's row group, as the section draws it. */
export interface QuestionBindingGroup {
  /** Whether this question is asked at all (#715), or `null` when the
   * switches have not been read yet.
   *
   * `null` rather than a defaulted `true`, on `TaskState.questionSwitches`'
   * own contract: a roster that drew ten toggles from a list it had not read
   * would state a fact about the workspace, and the first one to flip on the
   * next answer would look like a bug rather than an answer arriving. */
  enabled: boolean | null;
  /** Whether an unconfirmed toggle write is overlaid on this question — the
   * same read-time fact `BindingDTO.pending` carries for a binding row. */
  pending: boolean;
  /** The wire spelling of the question — a stable key for React, never
   * shown. */
  question: string;
  label: string;
  surface: string;
  /** The rows that answer this question, in the roster's own key order.
   * Empty for most questions, and an empty group is still rendered: a
   * question nobody has to configure is a fact, not an omission. */
  rows: BindingDTO[];
  /** Keys the roster says answer this question for which the reader was
   * given no row at all.
   *
   * `Core::bindings` returns every key it knows, set or not, so in
   * production this is always empty — but the demo world hand-authors a
   * subset, and reporting a question with a *missing* row as a question
   * with *nothing to set* would be the flat opposite of true. The two
   * cases are separated here so the screen can say which one it is. */
  missing: string[];
}

/** The section's whole shape: every question, then whatever `settings` rows
 * belonged to none of them. */
export interface GroupedBindings {
  groups: QuestionBindingGroup[];
  /** Live rows no question claims — in practice the keys this build cannot
   * write (`BindingDTO.known === false`), which `Core::bindings` returns on
   * purpose so the editor shows what is really in the table. Dropping them
   * here would be the regression `bindings.rs` warns about. */
  other: BindingDTO[];
}

/** Folds `Core::bindings`' flat answer into the core's question roster.
 *
 * Never invents or hides a row: every input row lands in exactly one group
 * or in `other`, and the count is asserted by this module's own test. */
export function groupBindingsByQuestion(
  roster: readonly QuestionRosterEntry[],
  bindings: readonly BindingDTO[],
  /** `Core::question_switches`' answer (#715), or `null` before the first
   * one arrives. A question with no entry in a non-null list is `null` too —
   * "this reader was handed no switch for it", which is the same distinction
   * `missing` draws for a binding row and is exactly the demo-world
   * asymmetry that shipped a copy bug at #714. */
  switches: readonly QuestionSwitchDTO[] | null = null,
): GroupedBindings {
  const claimed = new Set<string>();
  const groups = roster.map((entry) => {
    const rows: BindingDTO[] = [];
    const missing: string[] = [];
    for (const key of entry.bindings) {
      const row = bindings.find((binding) => binding.key === key);
      if (row === undefined) {
        missing.push(key);
        continue;
      }
      claimed.add(key);
      rows.push(row);
    }
    const switchState = switches?.find((candidate) => candidate.question === entry.question);
    return {
      question: entry.question,
      label: entry.label,
      surface: entry.surface,
      enabled: switchState === undefined ? null : switchState.enabled,
      pending: switchState?.pending ?? false,
      rows,
      missing,
    };
  });
  return { groups, other: bindings.filter((binding) => !claimed.has(binding.key)) };
}

/** The message a failed toggle write should show on its own question's row,
 * or `null` when the last write was another question's or succeeded —
 * [`bindingWriteError`] verbatim for the question vocabulary, matched by
 * question so a stale failure never bleeds onto a row it did not come
 * from. */
export function questionSwitchWriteError(
  lastWrite: TaskQuestionSwitchResult | null,
  question: string,
): string | null {
  if (lastWrite === null || lastWrite.question !== question || lastWrite.kind === "ok") {
    return null;
  }
  switch (lastWrite.kind) {
    case "unknown_question":
      return "This build doesn't know that question, so it wasn't switched.";
    case "busy":
      return "The core was busy. Try again.";
    case "failed":
      return lastWrite.error ?? "That switch didn't save.";
  }
}

/** Which question a binding key answers, in the reader's words — the lookup
 * that replaced the calendar picker's hand-written *"it answers How long to
 * the next vacation"* (#714). `null` when no question claims the key, which
 * is what an unwritable row is; the caller says something else then rather
 * than naming a question that does not exist. */
export function questionLabelForBinding(
  roster: readonly QuestionRosterEntry[],
  key: string,
): string | null {
  const owner = roster.find((entry) => entry.bindings.includes(key));
  return owner === undefined ? null : owner.label;
}

File: client/core/src/bindings.rs
//! The standing-question **bindings** (#118, ADR-0015): the small
//! cross-device facts a pane needs before it can answer anything — which
//! race series, which calendar holds the trips, which page the waste
//! schedule is scraped from — and, since the homework pane's standing link,
//! one fact that names no source at all ([`BindingKey::HomeworkLink`]): a
//! value that must not live in a public repo, held here for the same
//! sync-it-once reason rather than because a pane queries it.
//!
//! A binding is one `settings` row (ADR-0009's KV table), so it rides the
//! ordinary delta pull and the full sweep like every other table, and is
//! written as an ordinary entity-level CAS set. There is no bespoke sync
//! path here and there must never be one: `SyncMirror` already carries
//! `settings`, and [`crate::Core::set_binding`] enqueues through
//! [`crate::sync::SyncCycle::enqueue`] exactly as `capture`/`act`/`triage`
//! do.
//!
//! **Keys are kebab-case and unversioned** (ADR-0015, decided in the seam
//! grilling that opened this slice). Deliberately unversioned: a source
//! bump — `city-waste/v1` → `city-waste/v2` — must not orphan a binding in
//! a table that has no DELETE, and a key carrying a version string would do
//! exactly that. The *source* is versioned; the question the human answered
//! once ("which page is my collection schedule on") is not.
//!
//! **The vocabulary is closed** ([`BindingKey`]), and that is the point of
//! this module: `settings` is a generic KV table with no DELETE, so a UI
//! able to mint arbitrary keys can only ever grow it. Every key crossing a
//! seam is resolved by name through [`BindingKey::parse`] before it reaches
//! `Core`, the same discipline `ffi-web`'s `parse_action`/`parse_destination`
//! already apply to their own wire strings. A key this build has never
//! heard of is still *readable* — see [`Binding::known`] — because a row
//! written by a newer build is a fact about this build, not a broken row
//! (the same reading ADR-0015 gives an unrecognised snapshot `schema`).
//!
//! **The collapse state of a pane is not a binding** (ADR-0015, again
//! explicitly): it is device-local and band-scoped, and belongs in the
//! injectable-`storage` idiom `calendar/persistence.ts` / `theme/useTheme.ts`
//! already use. Nothing in this module is a view preference.

use serde::Serialize;

/// Every binding this build knows how to write. Closed, kebab-case,
/// unversioned — see the module docs for all three reasons.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum BindingKey {
    /// Which motorsport series the next-race pane follows (#119).
    RaceSeries,
    /// The dedicated Trips calendar the vacation countdown reads (#121) —
    /// a calendar id, since the calendar stays the authority for its own
    /// events (ADR-0005).
    TripsCalendar,
    /// The council page the waste schedule is polled from (#120).
    CityWastePage,
    /// The standing meeting link the homework pane offers (#675's
    /// follow-on).
    ///
    /// **Unlike its three siblings this names no source.** `race-series`,
    /// `trips-calendar` and `city-waste-page` each say *which thing a pane
    /// reads*, and an unset one leaves that pane `AnswerState::Unbound`
    /// with a setup prompt. This one names a **display affordance** the
    /// homework pane renders and never queries: unset, the pane answers
    /// exactly as it does today and simply draws no button (ADR-0015,
    /// amended).
    ///
    /// It is a binding rather than a literal beside `HOMEWORK_CONTEXT` for
    /// one reason: the value is a meeting URL with its passcode embedded in
    /// it, and this repo is public. A binding is one `settings` row on the
    /// operator's own authority, set once and synced to every device —
    /// never in git.
    HomeworkLink,
}

impl BindingKey {
    /// Every key, in the order [`crate::Core::bindings`] presents them —
    /// declaration order, not alphabetical, so the editor's rows stay put
    /// as the vocabulary grows.
    pub const ALL: &'static [BindingKey] = &[
        BindingKey::RaceSeries,
        BindingKey::TripsCalendar,
        BindingKey::CityWastePage,
        BindingKey::HomeworkLink,
    ];

    /// The `settings.key` this binding is stored under — the wire spelling
    /// in both directions.
    pub fn as_str(self) -> &'static str {
        match self {
            BindingKey::RaceSeries => "race-series",
            BindingKey::TripsCalendar => "trips-calendar",
            BindingKey::CityWastePage => "city-waste-page",
            BindingKey::HomeworkLink => "homework-link",
        }
    }

    /// Resolves a wire key by name. `None` is a key this build does not
    /// know how to *write* — never a reason to hide a row it can read.
    pub fn parse(key: &str) -> Option<BindingKey> {
        BindingKey::ALL
            .iter()
            .copied()
            .find(|candidate| candidate.as_str() == key)
    }
}

/// What a binding is currently set to, as three distinct states rather than
/// a nullable string.
///
/// The split exists for the same reason [`crate::freshness::Freshness`]
/// refuses to be a boolean: "nobody has set this yet" and "this holds
/// something that is not text" are different facts, and collapsing the
/// second into the first would let the editor silently offer to overwrite a
/// value it never showed anyone. [`BindingValue::Other`] carries the stored
/// JSON verbatim so a reader can put it on screen — visibly odd, never
/// quietly empty (ADR-0015).
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(tag = "state", rename_all = "snake_case")]
pub enum BindingValue {
    /// No `settings` row for this key (or the row has gone absent —
    /// ADR-0003's demotion, which reads the same way here: this device has
    /// no current answer).
    Unset,
    /// The stored JSON is a string — the shape every binding this slice
    /// writes uses, and the only one [`crate::Core::set_binding`] produces.
    Text { text: String },
    /// The stored JSON is something else (a number, an object, a value a
    /// newer build wrote). Carried as its canonical JSON text so it can be
    /// displayed; deliberately not parsed further and not editable here.
    Other { raw: String },
}

impl BindingValue {
    /// Reads one stored `settings.value` — canonical JSON text, per
    /// `hummingbird_domain::Setting`'s own doc — into the three states
    /// above. Unparseable JSON lands in [`BindingValue::Other`] with the
    /// bytes as stored rather than being dropped: the row exists, and
    /// saying nothing about it would be the quietly-empty answer.
    pub fn from_stored(raw: &str) -> BindingValue {
        match serde_json::from_str::<serde_json::Value>(raw) {
            Ok(serde_json::Value::String(text)) => BindingValue::Text { text },
            _ => BindingValue::Other {
                raw: raw.to_string(),
            },
        }
    }
}

/// One binding as a reader sees it: every [`BindingKey`] this build knows
/// (set or not), plus every other live `settings` row, so the editor can
/// show what is actually in the table rather than only what it can write.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Binding {
    pub key: String,
    /// Whether [`BindingKey::parse`] resolved this key — i.e. whether this
    /// build can write it. A `false` row is display-only.
    pub known: bool,
    /// Whether an unconfirmed local write is currently overlaid on this key
    /// — the same read-time overlay fact `Core::is_pending` reports for an
    /// item, never a stored column.
    pub pending: bool,
    pub value: BindingValue,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_key_is_kebab_case_and_carries_no_version_suffix() {
        // ADR-0015: a `/vN` in a binding key would orphan the binding on a
        // source bump, in a table with no DELETE.
        for key in BindingKey::ALL {
            let spelling = key.as_str();
            assert!(
                spelling
                    .chars()
                    .all(|c| c.is_ascii_lowercase() || c == '-'),
                "{spelling} must be kebab-case"
            );
            assert!(!spelling.contains('/'), "{spelling} must carry no version");
            assert!(
                !spelling.chars().any(|c| c.is_ascii_digit()),
                "{spelling} must carry no version"
            );
        }
    }

    #[test]
    fn the_vocabulary_is_the_three_adr_0015_sources_then_the_homework_link() {
        // Declaration order is the editor's row order, so the fourth is
        // APPENDED — the three source bindings keep their rows. It is also
        // deliberately last in a second sense: it is not a source binding
        // at all (see `BindingKey::HomeworkLink`), so nothing above it
        // becomes harder to read by its arrival.
        let spellings: Vec<&str> = BindingKey::ALL.iter().map(|key| key.as_str()).collect();
        assert_eq!(
            spellings,
            vec![
                "race-series",
                "trips-calendar",
                "city-waste-page",
                "homework-link"
            ]
        );
    }

    #[test]
    fn a_key_round_trips_by_name_and_an_unknown_one_is_rejected() {
        for key in BindingKey::ALL {
            assert_eq!(BindingKey::parse(key.as_str()), Some(*key));
        }
        assert_eq!(BindingKey::parse("race-series/v1"), None);
        assert_eq!(BindingKey::parse("raceSeries"), None);
        assert_eq!(BindingKey::parse(""), None);
    }

    #[test]
    fn a_stored_value_reads_as_text_only_when_it_actually_is_one() {
        assert_eq!(
            BindingValue::from_stored("\"f1\""),
            BindingValue::Text {
                text: "f1".to_string()
            }
        );
        // A number, an object and outright garbage are all "something else
        // is in there", shown verbatim rather than reported as unset.
        for raw in ["7", "{\"a\":1}", "null", "not json"] {
            assert_eq!(
                BindingValue::from_stored(raw),
                BindingValue::Other {
                    raw: raw.to_string()
                },
                "{raw}"
            );
        }
    }

    #[test]
    fn the_value_states_serialize_as_a_tagged_union_never_a_nullable_string() {
        assert_eq!(
            serde_json::to_string(&BindingValue::Unset).unwrap(),
            r#"{"state":"unset"}"#
        );
        assert_eq!(
            serde_json::to_string(&BindingValue::Text {
                text: "f1".to_string()
            })
            .unwrap(),
            r#"{"state":"text","text":"f1"}"#
        );
        assert_eq!(
            serde_json::to_string(&BindingValue::Other {
                raw: "7".to_string()
            })
            .unwrap(),
            r#"{"state":"other","raw":"7"}"#
        );
    }
}
